parser/src/cursor.rs

```diff
@@ -0,0 +1,88 @@
+use crate::moves::Move;
+use lexer::token::Token;
+use lexer::token::TokenType::EndOfFile;
+
+use crate::parser::{ParseError, ParseResult};
+
+/// Parser cursor is used by parsers to navigate in the token stream
+#[derive(Debug, Clone)]
+pub(crate) struct ParserCursor<'a> {
+    /// The manipulated tokens
+    tokens: Vec<Token<'a>>,
+    /// current position in the tokens vector.
+    pos: usize,
+}
+
+impl<'a> ParserCursor<'a> {
+    ///Creates a new cursor at position 0 in the given token vector
+    pub fn new(tokens: Vec<Token<'a>>) -> Self {
+        Self { tokens, pos: 0 }
+    }
+
+    ///advance if next token satisfy the given move
+    /// Returns the token where the move ended if it succeeds, or None instead.
+    /// This method will move the current cursor position on where the move ended.
+    pub fn advance(&mut self, mov: impl Move) -> Option<Token<'a>> {
+        let result = mov.apply(|pos| self.at(pos), self.pos);
+
+        if let Some(new_pos) = result {
+            self.pos = new_pos;
+            return Some(self.at(new_pos - 1));
+        }
+        None
+    }
+
+    /// Returns the token where the move ended if the move succeeds, or None instead.
+    /// This method is similar to `advance` except that it does not makes the cursor change its current pos.
+    pub fn lookahead(&self, mov: impl Move) -> Option<Token<'a>> {
+        let result = mov.apply(|pos| self.at(pos), self.pos);
+
+        if let Some(new_pos) = result {
+            return Some(self.at(new_pos));
+        }
+        None
+    }
+
+    /// Force the given move to succeed, or else fail with given error message.
+    /// This method will move the current cursor position on where the move ended.
+    pub fn force(&mut self, mov: impl Move, err: &str) -> ParseResult<Token<'a>> {
+        self.advance(mov).ok_or_else(|| ParseError {
+            message: err.to_string(),
+        })
+    }
+
+    ///returns the token at current position
+    pub fn peek(&self) -> Token<'a> {
+        self.at(self.pos)
+    }
+
+    ///returns current token then advance or ParseError if this cursor hits the
+    /// end of the stream.
+    pub fn next(&mut self) -> ParseResult<Token<'a>> {
+        self.next_opt().ok_or(ParseError {
+            message: "Unexpected end of file".to_string(),
+        })
```

Comment:
```suggestion
        self.next_opt().ok_or_else(|| ParseError {
            message: "Unexpected end of file".to_string(),
        })
```
The string don't need to be allocated every time `next()` is called.